ray-operator/test/e2erayservice/rayservice_in_place_update_test.go

```diff
@@ -82,3 +86,177 @@ func TestRayServiceInPlaceUpdate(t *testing.T) {
+	curlPodName := "curl-pod"
+	curlContainerName := "curl-container"
+
+	curlPod, err := CreateCurlPod(test, curlPodName, curlContainerName, namespace.Name)
+	g.Expect(err).NotTo(HaveOccurred())
+	// Wait until curl pod is created
+	g.Eventually(func(g Gomega) *corev1.Pod {
+		updatedCurlPod, err := test.Client().Core().CoreV1().Pods(curlPod.Namespace).Get(test.Ctx(), curlPod.Name, metav1.GetOptions{})
+		g.Expect(err).NotTo(HaveOccurred())
+		return updatedCurlPod
+	}, TestTimeoutShort).Should(WithTransform(sampleyaml.IsPodRunningAndReady, BeTrue()))
+
+	LogWithTimestamp(test.T(), "Sending requests to the RayService to make sure it is ready to serve requests")
+	stdout, _ := CurlRayServicePod(test, rayService, curlPod, curlContainerName, "/fruit", `["MANGO", 2]`)
+	g.Expect(stdout.String()).To(Equal("6"))
+	stdout, _ = CurlRayServicePod(test, rayService, curlPod, curlContainerName, "/calc", `["MUL", 3]`)
+	g.Expect(stdout.String()).To(Equal("15 pizzas please!"))
+
+	// In-place update
+	// Parse ServeConfigV2 and replace the string in the simplest way to update it.
+	rayService, err = GetRayService(test, namespace.Name, rayService.Name)
+	g.Expect(err).NotTo(HaveOccurred())
+
+	serveConfig := rayService.Spec.ServeConfigV2
+	serveConfig = strings.Replace(serveConfig, "price: 3", "price: 456", -1)
+	rayService.Spec.ServeConfigV2 = serveConfig
+
+	rayService.Spec.RayClusterSpec.HeadGroupSpec.Template.Spec.Containers[0].Env = []corev1.EnvVar{{Name: "FOO", Value: "BAR"}}
+
+	var activeRayClusterBeforeUpdate *rayv1.RayCluster
+	activeRayClusterBeforeUpdate, err = GetRayCluster(test, namespace.Name, rayService.Status.ActiveServiceStatus.RayClusterName)
+	g.Expect(err).NotTo(HaveOccurred())
+
+	rayService, err = test.Client().Ray().RayV1().RayServices(namespace.Name).Update(
+		test.Ctx(),
+		rayService,
+		metav1.UpdateOptions{},
+	)
+	g.Expect(err).NotTo(HaveOccurred())
+
+	// Wait active RayCluster has been terminated
+	g.Eventually(func() bool {
+		_, err := GetRayCluster(test, activeRayClusterBeforeUpdate.Namespace, activeRayClusterBeforeUpdate.Name)
+		return k8sApiErrors.IsNotFound(err)
+	}, TestTimeoutMedium).Should(BeTrue())
+
+	// Make sure the serveConfig is updated to new ray cluster.
+	g.Eventually(func(g Gomega) {
+		// curl /fruit
+		stdout, _ := CurlRayServicePod(test, rayService, curlPod, curlContainerName, "/fruit", `["MANGO", 2]`)
+		g.Expect(stdout.String()).To(Equal("912"))
+	}, TestTimeoutShort).Should(Succeed())
+
+	events, err := test.Client().Core().CoreV1().Events(rayService.Namespace).List(test.Ctx(), metav1.ListOptions{
+		FieldSelector: fields.OneTermEqualSelector("involvedObject.name", rayService.Name).String(),
+	})
+	g.Expect(err).NotTo(HaveOccurred())
+
+	// event UpdatedServeApplications should occure exactly twice during the entire test.
+	count := 0
+	for _, event := range events.Items {
+		if event.Reason != string(utils.UpdatedServeApplications) {
+			continue
+		}
+		count++
+	}
+	g.Expect(count).To(Equal(2))
```

Comment:
Can you also check these two events? One is for the old RayCluster, and the other is for the new RayCluster.

Reply:
The RayCluster name appears on `message` which might be exclude from aggregation. It would be difficult to distinguish which is which. There are some discussion about this condition https://github.com/ray-project/kuberay/pull/3559#discussion_r2084908259. Or do I misunderstand something?

Reply:
In this case, why can we check the number of events?

Reply:
Discuss offline:

I am OK to merge the PR if:

0. Survey when will aggregation be triggered.
1. It can pass 20 times consecutively in your local env.
2. Check these two events are for different RayCluster CRs
3. Add a comment / TODO for the possibility of aggregation

Reply:
https://gist.github.com/rueian/fca6c2597ff2ba1dfc2ab5eee148bee7

Reply:
0. `Aggregation is performed if a similar event is recorded 10 times in a 10 minute rolling interval.`
https://github.com/kubernetes/client-go/blob/46965213e4561ad1b9c585d1c3551a0cc8d3fcd6/tools/record/events_cache.go#L430-L432

1. Here is the screenshot
<img width="647" alt="rayservice e2e" src="https://github.com/user-attachments/assets/583a7a56-7015-4c0e-ab05-bffd9b8a91ab" />

2. Compare two RayClusters name from the event message at the end.

Reply:
> Aggregation is performed if a similar event is recorded 10 times in a 10 minute rolling interval.
https://github.com/kubernetes/client-go/blob/46965213e4561ad1b9c585d1c3551a0cc8d3fcd6/tools/record/events_cache.go#L430-L432

Oh, sorry. In that case, aggregation is quite likely to happen since we share the Kubernetes cluster across multiple end-to-end tests. We might need an alternative way to verify that the Ray Serve applications in the old RayCluster are not being updated.

A possible way is:

1. Add a finalizer to old RayCluster
2. Trigger zero downtime upgrade
3. Wait until RayService switches to the new RayCluster
4. Verify the old RayCluster's Serve applications are not updated.
5. Remove the finalizer

Reply:
Fixed. PTAL